Purchases/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift

```diff
@@ -0,0 +1,149 @@
+@available(iOS 12.2, macOS 10.14.4, tvOS 12.2, watchOS 6.2, *)
+public typealias SK1ProductDiscount = SKProductDiscount
+
+/// TypeAlias to StoreKit 2's Discount type, called `StoreKit.Product.SubscriptionOffer`
+@available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+public typealias SK2ProductDiscount = StoreKit.Product.SubscriptionOffer
+
+@objc(RCStoreProductDiscount)
+public final class StoreProductDiscount: NSObject, StoreProductDiscountType {
+
+    @objc(RCPaymentMode)
+    public enum PaymentMode: Int {
+
+        case none = -1
+        case payAsYouGo = 0
+        case payUpFront = 1
+        case freeTrial = 2
+
+    }
+
+    private let discount: StoreProductDiscountType
+
+    init(_ discount: StoreProductDiscountType) {
+        self.discount = discount
+
+        super.init()
+    }
+
+    @objc public var offerIdentifier: String? { self.discount.offerIdentifier }
+    @objc public var price: Decimal { self.discount.price }
+    @objc public var paymentMode: PaymentMode { self.discount.paymentMode }
+    @objc public var subscriptionPeriod: SubscriptionPeriod { self.discount.subscriptionPeriod }
+
+    /// Creates an instance from any `StoreProductDiscountType`.
+    /// If `discount` is already a wrapped `StoreProductDiscount` then this returns it instead.
+    static func from(discount: StoreProductDiscountType) -> StoreProductDiscount {
+        return discount as? StoreProductDiscount
+        ?? StoreProductDiscount(discount)
+    }
+
+    public override func isEqual(_ object: Any?) -> Bool {
+        guard let other = object as? StoreProductDiscount else { return false }
+
+        return self.offerIdentifier == other.offerIdentifier
+            && self.price == other.price
+            && self.paymentMode == other.paymentMode
+            && self.subscriptionPeriod == other.subscriptionPeriod
+    }
+
+    public override var hash: Int {
+        var hasher = Hasher()
+        hasher.combine(self.offerIdentifier)
+        hasher.combine(self.price)
+        hasher.combine(self.paymentMode)
+        hasher.combine(self.subscriptionPeriod)
+
+        return hasher.finalize()
+    }
+
+}
+
+/// The details of an introductory offer or a promotional offer for an auto-renewable subscription.
+internal protocol StoreProductDiscountType {
+
+    /// A string used to uniquely identify a discount offer for a product.
+    var offerIdentifier: String? { get }
+
+    /// The discount price of the product in the local currency.
+    var price: Decimal { get }
+
+    /// The payment mode for this product discount.
+    var paymentMode: StoreProductDiscount.PaymentMode { get }
+
+    /// The period for the product discount.
+    var subscriptionPeriod: SubscriptionPeriod { get }
+
+}
+
+// MARK: - Wrapper constructors / getters
+
+extension StoreProductDiscount {
+
+    @available(iOS 12.2, macOS 10.14.4, tvOS 12.2, watchOS 6.2, *)
+    internal convenience init(sk1Discount: SK1ProductDiscount) {
```

Comment:
we can skip `internal` for these, but feel free to keep them if you wanted to be very explicit since there are `public` methods in the same extension

Reply:
Yeah I mainly did it to document that we don't actually want to expose these, so it makes it explicit that this is a conscious decision.